pkg/algorithm/sku_scheduler.go

```diff
@@ -0,0 +1,517 @@
+		isAllocated := false
+		for _, c := range allocatedCells {
+			if isAncestor(cell, c) || isAncestor(c, cell) {
+				isAllocated = true
+				break
+			}
+		}
+		if !isAllocated {
+			if cell.GetPriority() == freePriority {
+				freeCells = append(freeCells, cell)
+			} else if cell.GetPriority() < priority {
+				preemptibleCells = append(preemptibleCells, cell)
+			}
+		}
+	}
+	return freeCells, preemptibleCells
+}
+
+// checkOptimalAffinityForCells checks if the currently picked cells have the lowest LCA.
+// It also checks if the solution is optimal (if the leaf cells are all buddies).
+func checkOptimalAffinityForCells(
+	affinity CellLevel,
+	freeCells CellList,
+	currentCellIndices []int32,
+	bestAffinity CellLevel,
+	bestAffinityCells CellList,
+	bestAffinityCellIndices []int32,
+	optimalAffinity CellLevel) (CellLevel, bool) {
+
+	if affinity < bestAffinity {
+		copy(bestAffinityCellIndices, currentCellIndices)
+		for i := 0; i < len(currentCellIndices); i++ {
+			bestAffinityCells[i] = freeCells[currentCellIndices[i]]
+		}
+		if affinity == optimalAffinity {
+			return affinity, true
+		} else {
+			return affinity, false
+		}
+	}
+	return bestAffinity, false
+}
+
+func (s *skuScheduler) createSkuClusterView(
+	within *skuCell,
+	withinLevel CellLevel,
+	priority CellPriority) skuClusterView {
+
+	cv := skuClusterView{}
+	for l := withinLevel; l >= CellLevel(1); l-- {
+		for _, c := range s.chainCellList[l] {
+			if (within != nil && !isAncestor(within.cell, c)) ||
+				cv.contains(ancestorNoHigherThanLevel(withinLevel, c)) {
+				continue
+			}
+			cell := &skuCell{
+				cell:                          c,
+				freeLeafCellNumAtPriority:     c.GetTotalLeafCellNum(),
+				usedLeafCellNumAtPriority:     0,
+				usedLeafCellNumHigherPriority: 0,
+				healthy:                       true,
+				address:                       "",
+			}
+			for p, num := range c.GetUsedLeafCellNumAtPriorities() {
+				if p >= priority {
+					cell.freeLeafCellNumAtPriority -= num
+				}
+				if s.crossPriorityPack {
+					cell.usedLeafCellNumAtPriority += num
+				} else {
+					if p == priority {
+						cell.usedLeafCellNumAtPriority += num
+					}
+					if p > priority {
+						cell.usedLeafCellNumHigherPriority += num
+					}
+				}
+			}
+			switch v := c.(type) {
+			case *PhysicalCell:
+				cell.healthy = v.IsHealthy()
+				cell.address = c.GetAddress()
+			case *VirtualCell:
+				if pn := v.GetPhysicalCell(); pn != nil {
+					cell.healthy = pn.IsHealthy()
+					cell.address = pn.GetAddress()
+				}
+			}
+			cv = append(cv, cell)
+		}
+	}
+	sort.Stable(cv)
+	return cv
+}
+
+// Len method for sorting sku cells in cluster view
+func (cv skuClusterView) Len() int {
+	return len(cv)
+}
+
+// Less method for sorting sku cells in cluster view
+// sort in the following order:
+// 1. cell health (prefer healthy)
+// 2. cell level (prefer lower)
+// 3. usedLeafCellNumAtPriority (prefer higher)
+// 4. usedLeafCellNumHigherPriority (prefer lower)
```

Comment:
if greedy search, may need use same order to sort here?

Reply:
Comment usedLeafCellNumAtPriority different meaning in different crossPriorityPack case

Reply:
updated